### PR TITLE
CTM-611: added logic for batch run error handling.

### PR DESCRIPTION
**Pull request by kevinwangcanada**

### CODE REVIEW

File: matchminer/message/rabbitmq_message.py [hunk 1]
```diff
@@ -117,11 +117,52 @@ def process_job(self, ch, method, properties, body):
...
+                        "user_id": user_id,
+                        "trial_internal_ids": trial_internal_ids,
+                        "run_status": "success",
+                        "run_message": success_msg,
+                        "failed_trial_internal_ids": failed_trial_internal_ids,
+                    }
+                    json_success_msg = json.dumps(py_success_dict)
+                    logging.info(json_success_msg)
+                    self.send_message(json_success_msg)
+                if(num_failed_trials == num_trials):
```
Comment (mickey-ng) [inline matchminer/message/rabbitmq_message.py]: use elif (else if)

File: matchminer/message/rabbitmq_message.py [hunk 2]
```diff
@@ -117,11 +117,52 @@ def process_job(self, ch, method, properties, body):
...
+                        "user_id": user_id,
+                        "trial_internal_ids": trial_internal_ids,
+                        "run_status": "fail",
+                        "run_message": error_msg,
+                        "failed_trial_internal_ids": failed_trial_internal_ids,
+                    }
+                    json_error_msg = json.dumps(py_error_dict)
+                    logging.error(json_error_msg)
+                    self.send_message(json_error_msg)
+                if(num_failed_trials < num_trials and num_failed_trials >= 1):
```
Comment (mickey-ng) [inline matchminer/message/rabbitmq_message.py]: use 'else', I don't think there are other cases we need to consider

File: matchminer/custom.py [hunk 3]
```diff
@@ -1423,12 +1424,13 @@ def run_ctims_matchengine_job(trial_internal_ids):
...
             ignore_run_log=True,
             ignore_report_date=True,
-            protocol_nos=trial_internal_ids
+            protocol_nos=trial_internal_ids,
+            failed_protocol_nos={}
     ) as me:
         me.get_matches_for_all_trials()
-        me.update_all_matches()
+        if (len(me.failed_protocol_nos.keys()) > 0):
+            return me.failed_protocol_nos
```
Comment (mickey-ng) [inline matchminer/custom.py]: what happens to the partial success? It won't update_all_matches() for partial success ones

File: matchminer/custom.py [hunk 4]
```diff
@@ -1423,12 +1424,13 @@ def run_ctims_matchengine_job(trial_internal_ids):
             db_name='matchminer',
             ignore_run_log=True,
             ignore_report_date=True,
-            protocol_nos=trial_internal_ids
+            protocol_nos=trial_internal_ids,
+            failed_protocol_nos={}
```
Comment (mickey-ng) [inline matchminer/custom.py]: failed_protocol_nos isn't a configurable value, and it's defined in the engine itself, so no need to pass this information to the PMatchEngine

File: matchminer/message/rabbitmq_message.py (hunk 1 above)
Comment (kevinwangcanada) [reply matchminer/message/rabbitmq_message.py]: ok

File: matchminer/message/rabbitmq_message.py (hunk 2 above)
Comment (kevinwangcanada) [reply matchminer/message/rabbitmq_message.py]: ok

File: matchminer/custom.py (hunk 3 above)
Comment (kevinwangcanada) [reply matchminer/custom.py]: good catch.

File: matchminer/custom.py (hunk 4 above)
Comment (kevinwangcanada) [reply matchminer/custom.py]: ok

File: matchminer/custom.py [hunk 5]
```diff
@@ -1413,6 +1413,10 @@ def run_ctims_matchengine_job(trial_internal_ids):
     installed_dir = sys.prefix
     plugin_dir = os.path.join(installed_dir, 'pugh-lab')
     file_dir = os.path.join(plugin_dir, 'config.json')
+    # plugin_dir = os.path.join(installed_dir, '../../pmatchengine/pugh-lab/plugins')
+    # config_dir = os.path.join(installed_dir, '../../pmatchengine/pugh-lab')
+    # file_dir = os.path.join(config_dir, 'config.json')
```
Comment (mickey-ng) [inline matchminer/custom.py]: can you remove these lines?

File: matchminer/custom.py [hunk 6]
```diff
@@ -1427,8 +1431,7 @@ def run_ctims_matchengine_job(trial_internal_ids):
...
         me.get_matches_for_all_trials()
         me.update_all_matches()
+        if (len(me.failed_protocol_nos.keys()) > 0):
+            return me.failed_protocol_nos
 
-    resp = Response(response=json.dumps({"success": True}),
-                    status=200,
-                    mimetype="application/json")
-    return resp
+    return failed_protocol_nos
```
Comment (mickey-ng) [inline matchminer/custom.py:1434]: This will return failed_protocol_nos twice?

File: matchminer/message/rabbitmq_message.py (hunk 2 above)
Comment (kevinwangcanada) [reply matchminer/message/rabbitmq_message.py]: removed

File: matchminer/custom.py (hunk 6 above)
Comment (kevinwangcanada) [reply matchminer/custom.py:1434]: I changed the code.

**Review by mickey-ng — APPROVED** [review]

**Pull request merged**